src/libopensc/card-piv.c

```diff
@@ -467,82 +791,1819 @@ static struct sc_card_driver piv_drv = {
+
+	/* If already read and matches previous version return SC_SUCCESS */
+	if (cvc->der.value && (cvc->der.len == *buflen) && buf && *buf
+			&& (memcmp(cvc->der.value, *buf, *buflen) == 0))
+		LOG_FUNC_RETURN(card->ctx, SC_SUCCESS);
+
+	piv_clear_cvc_content(cvc);
+
+	memset(cvc, 0, sizeof(piv_cvc_t));
+	cvc->issuerIDlen = sizeof(cvc->issuerID);
+	cvc->subjectIDlen = sizeof(cvc->subjectID);
+
+	sc_copy_asn1_entry(c_asn1_piv_cvc, asn1_piv_cvc);
+	sc_copy_asn1_entry(c_asn1_piv_cvc_body, asn1_piv_cvc_body);
+	sc_copy_asn1_entry(c_asn1_piv_cvc_pubkey, asn1_piv_cvc_pubkey);
+	sc_copy_asn1_entry(c_asn1_piv_cvc_dsobj, asn1_piv_cvc_dsobj);
+	sc_copy_asn1_entry(c_asn1_piv_cvc_dssig, asn1_piv_cvc_dssig);
+	sc_copy_asn1_entry(c_asn1_piv_cvc_alg_id, asn1_piv_cvc_alg_id);
+
+	sc_format_asn1_entry(asn1_piv_cvc_alg_id    , &cvc->signatureAlgOID, NULL, 1);
+	sc_format_asn1_entry(asn1_piv_cvc_alg_id + 1, NULL, NULL, 1); /* NULL */
+
+	sc_format_asn1_entry(asn1_piv_cvc_dssig    , &asn1_piv_cvc_alg_id, NULL, 1);
+	sc_format_asn1_entry(asn1_piv_cvc_dssig + 1, &cvc->signature, &signaturebits, 1);
+
+	sc_format_asn1_entry(asn1_piv_cvc_dsobj    , &asn1_piv_cvc_dssig, NULL, 1);
+
+	sc_format_asn1_entry(asn1_piv_cvc_pubkey    , &cvc->pubKeyOID, NULL, 1);
+	sc_format_asn1_entry(asn1_piv_cvc_pubkey + 1, &cvc->publicPoint, &cvc->publicPointlen, 1);
+
+	sc_format_asn1_entry(asn1_piv_cvc_body    , &cvc->cpi, NULL, 1);
+	sc_format_asn1_entry(asn1_piv_cvc_body + 1, &cvc->issuerID, &cvc->issuerIDlen, 1);
+	sc_format_asn1_entry(asn1_piv_cvc_body + 2, &cvc->subjectID, &cvc->subjectIDlen, 1);
+	sc_format_asn1_entry(asn1_piv_cvc_body + 3, &asn1_piv_cvc_pubkey, NULL, 1);
+	sc_format_asn1_entry(asn1_piv_cvc_body + 4, piv_get_asn1_obj, &roleIDder, 1);
+	sc_format_asn1_entry(asn1_piv_cvc_body + 5, &asn1_piv_cvc_dsobj, NULL, 1);
+
+	sc_format_asn1_entry(asn1_piv_cvc, &asn1_piv_cvc_body, NULL, 1);
+
+	r = sc_asn1_decode(card->ctx, asn1_piv_cvc, *buf, *buflen, NULL, NULL) ; /*(const u8 **) &buf_tmp, &len);*/
+	LOG_TEST_RET(card->ctx, r, "Could not decode card verifiable certificate");
+
+	cvc->signaturelen = signaturebits / 8;
+
+	if (roleIDder.len != 1)
+		LOG_TEST_RET(card->ctx, SC_ERROR_SM_AUTHENTICATION_FAILED, "roleID wrong length");
+
+	cvc->roleID = *roleIDder.value;
+
+	/* save body der for verification */
+	buf_tmp = *buf;
+	r = sc_asn1_read_tag(&buf_tmp, *buflen, &cla_out, &tag_out, &taglen);
+	LOG_TEST_RET(card->ctx, r," failed to read tag");
+
+	cvc->bodylen = (roleIDder.value + roleIDder.len) - buf_tmp;
+
+	cvc->body = malloc(cvc->bodylen);
+	if (cvc->body == NULL)
+		return SC_ERROR_OUT_OF_MEMORY;
+	memcpy(cvc->body, buf_tmp, cvc->bodylen);
+
+	/* save to reuse */
+	cvc->der.value = malloc(*buflen);
+	if (cvc->der.value == NULL) {
+		free(cvc->body);
+		return SC_ERROR_OUT_OF_MEMORY;
+	}
+	cvc->der.len = *buflen;
+	memcpy(cvc->der.value, *buf, cvc->der.len);
+
+	LOG_FUNC_RETURN(card->ctx, SC_SUCCESS);
+}
+#endif /* ENABLE_PIV_SM */
+
+#ifdef ENABLE_PIV_SM
```

Comment:
these ifdefs can be merged together. Not big deal, but makes less noise.

Reply:
Fixed.